copy/multiple.go

```diff
@@ -27,23 +27,35 @@ const (
 type instanceCopy struct {
 	op           instanceCopyKind
 	sourceDigest digest.Digest
+	requireCompressionFormatMatch bool
 }
 
 // prepareInstanceCopies prepares a list of instances which needs to copied to the manifest list.
-func prepareInstanceCopies(instanceDigests []digest.Digest, options *Options) []instanceCopy {
+func prepareInstanceCopies(list internalManifest.List, instanceDigests []digest.Digest, options *Options) ([]instanceCopy, error) {
 	res := []instanceCopy{}
 	for i, instanceDigest := range instanceDigests {
 		if options.ImageListSelection == CopySpecificImages &&
 			!slices.Contains(options.Instances, instanceDigest) {
 			logrus.Debugf("Skipping instance %s (%d/%d)", instanceDigest, i+1, len(instanceDigests))
 			continue
 		}
+		instanceDetails, err := list.Instance(instanceDigest)
+		if err != nil {
+			return res, err
+		}
+		requireCompressionFormatMatch := false
+		if options.DestinationCtx != nil && options.DestinationCtx.CompressionFormat != nil {
+			if !slices.Contains(instanceDetails.ReadOnly.CompressionAlgorithmNames, options.DestinationCtx.CompressionFormat.Name()) {
+				requireCompressionFormatMatch = true
+			}
+		}
```

Comment:
All of this about sending `require…Match` down to `copySingleImage` is relevant for the larger work, but not necessary for this PR, unless I’m missing something.

(And per #1987 , I default to thinking that it should be set only on the clone copies, but I’m not sure about that. … and there’s also the option for this PR to just build the infrastructure in `copySingleImage`, always setting it to `false`, and for #1987 to actually add the logic that sets it.)

---

If you intentionally want to make this PR add the two things (setting “update algorithms” correctly, and adding the `require…Match` infrastructure), that’s also a choice that can be made, just let me know whether you prefer this to do one or both.

Reply:
I think `require..Match` is needed for this case as well, consider the case where registry/infocache already contains blobs in a different compression and I am doing `--dest-format-compress` so afaik skopeo skips pushing blobs with `blob already exists` when `require..match` is not implemented (I assume it is reusing `gzip` blobs candidates). I am able to see this problem.

#### Without `require..match` does not copies new compression
```console
[fl@fedora bin]$ ./skopeo copy --all --format oci --dest-compress-format zstd --src-tls-verify=false --dest-tls-verify=false docker://localhost:5000/list docker://localhost:5000/list
Getting image list signatures
Copying 2 of 2 images in list
Copying image sha256:da39e651fdc155fbd53152dee387b91d4fd1e76da3137ccbd332f98e6c552a63 (1/2)
Getting image source signatures
Copying blob 8a117f3d2e73 skipped: already exists  
Copying blob 5119035169e2 skipped: already exists  
Copying config 58dce7cc03 done   | 
Writing manifest to image destination
Copying image sha256:f87412510ffdd07383a3b433ab58159522a4ffa3048f3557f55708ec90c8008c (2/2)
Getting image source signatures
Copying blob 33bbc77ac1e2 skipped: already exists  
Copying blob 5119035169e2 skipped: already exists  
Copying config 0044dbde8e done   | 
Writing manifest to image destination
Writing manifest list to image destination
Storing list signatures
``` 

#### With `require..match` correctly works
```console
[fl@fedora bin]$ ./skopeo copy --all --format oci --dest-compress-format zstd --src-tls-verify=false --dest-tls-verify=false docker://localhost:5000/list docker://localhost:5000/list
Getting image list signatures
Copying 2 of 2 images in list
Copying image sha256:da39e651fdc155fbd53152dee387b91d4fd1e76da3137ccbd332f98e6c552a63 (1/2)
Getting image source signatures
Copying blob 8a117f3d2e73 done   | 
Copying blob 5119035169e2 done   | 
Copying config 58dce7cc03 done   | 
Writing manifest to image destination
Copying image sha256:f87412510ffdd07383a3b433ab58159522a4ffa3048f3557f55708ec90c8008c (2/2)
Getting image source signatures
Copying blob 33bbc77ac1e2 done   | 
Copying blob 5119035169e2 done   | 
Copying config 0044dbde8e done   | 
Writing manifest to image destination
Writing manifest list to image destination
Storing list signatures
```

Reply:
I just added it cause it was constantly causing issues for me while testing locally, I suspect others might also hit this issue while trying to do the same, would it be an issue if we add it here and expose the option on frontend later in a different PR ? I think exposing part would be much simpler then.

Reply:
For the record, repeating from other discussion:
- The underlying `require…Match` functionality is definitely necessary for #1987 .
- I don’t, at this point, think that we should change the current behavior for `--dest-compress-format` (that’s https://github.com/containers/buildah/issues/4613 ); it should be a new option. I don’t feel _very_ strongly about this, and it might be worth discussing, but it also seems unrelated to the primary feature we are working towards at this very moment.

Reply:
Cool I have just implemented infra for now and passing `false` as default value, will implement this in a different PR.